### ISSUE TITLE
Akka streams doesn't send dropped messages to the dead letter office

### ISSUE THREAD

**Issue by jroper** [akka/akka#19222]
Sometimes you want to know when a message was dropped in Akka streams.  There are two use cases I can think of, one is debugging, which would be particularly useful for connections with bidirectional streams like sockets where terminating a stream in one direction implies termination in the other direction which loses messages.  The other is when messages need to be cleaned up, for example, you may be working with off heap memory (eg, when you're using Netty reactive streams, which publishes off heap backed messages), in these cases if a message is dropped, you need to release that memory.  In order to facilitate these use cases, Akka streams should make a best effort attempt to send all messages that it drops to the dead letter office.

**Comment by patriknw** [comment]
cc @rkuhn

**Comment by rkuhn** [comment]
@jroper Dropping elements in a stream is a completely normal and expected phenomenon, we offer a multitude of operators that perform this action. Stream cancellation or failure are other cases where elements will be dropped, and both of these are completely normal as well. Sending all dropped elements to some other destination is not even possible, think about allowing a user to supply the function that is going to be used in `map(...)`—this function might do horrible things!

What I am getting at is that stream elements cannot possibly have the expectation of being “kept live” or of having their lifecycle controlled while flowing through a stream, that is simply not possible. Sending pointers to off-heap memory into a stream is a no-go for this reason.

Another consideration is that producing large number of deadLetter messages is not only a performance concern, it also deeply concerns people to see anything show up there at all—it is not an adequate place where dropped elements should end up in.

But even if we set all this aside and assumed perfectly considered user-code that carefully retains all those off-heap pointers, sending these via Akka Actor messages will still not work because we do not (and cannot) guarantee delivery—which includes that some messages are lost even without going to deadLetters and there is nothing we can do about that.

In summary: thanks for creating this ticket, it will benefit future readers. We might also want to add this section to the docs.

**Comment by jroper** [comment]
Not every dropped message needs to go there - if you filter a stream, you don't need to send those dropped messages there, since you're intentionally dropping them.  It's just messages lost due to cancellation.  These shouldn't overwhelm the dead letter queue, it's no different from an actor going away because it's finished doing what it's doing, but the actor speaking to it hasn't received the message saying it's gone away yet, those messages go to the dead letter queue.

I understand that Akka doesn't guarantee delivery - but my understanding is that this is a blanket statement in response to people that want to ensure that messages aren't lost from a business perspective - eg, if the server crashes, network links go down etc, the messages can still be recovered. This level of guarantee is not needed for cleaning up memory, all the things that could prevent a message not arriving on the dead letter queue in Akka are the same things that could also prevent clean up when implementing handlers directly in Netty.  But cancelling a stream before it's consumed is a normal event that will gradually leak memory and always cause a server to crash if those messages aren't cleaned up - what's needed to counter this is not guaranteed delivery, but best effort.

**Comment by rkuhn** [comment]
What you are probably not aware of is that draining an Actor’s mailbox to deadLetters at the end of its lifecycle is indeed not reliable, this process routinely loses messages and fixing that would have a _huge_ performance impact on normal operations.

Hence I repeat: you cannot rely upon message delivery guarantees when it comes to required resource cleanup.

**Comment by rkuhn** [comment]
Another consideration is that Endre and myself spent many hours last week trying to figure out where in the HTTP codebase (unmarshallers etc.) some elements that were previously deemed discardable are dropped—because it turned out that they were not actually discardable after all. Saying that `filter` only does what the user tells it to may be technically correct, but it does not yield a sustainable solution.

**Comment by drewhk** [comment]
I want to add to this that we have discussed these issues very thoroghly back then when the first stream-of-stream operations were designed (dropping substreams) and at the end there was no solution to this. We are not doing a "blanket statement" here, this was debated fiercely multiple times before we reached our conclusion.

**Comment by jroper** [comment]
Stream of streams doesn't seem to be a problem if the

**Comment by drewhk** [comment]
But to not to sound handwaveiy, here are some issues among others:

 - what @rkuhn said about user code dropping messages. You might say that this is intentional then by the user, but it is not necessarily so in practice. 
 - 3rd party RS processors will probably not obey whatever cleanup strategy Akka would devise
 - 3rd party materializers might change the semantics of the cleanup strategy
 - piping messages through actors would probably also change the semantics

You might argue though that all the above is special cases, and still a resonable solution could be found for a subset (only Akka stages and materializer, user is sufficiently educated). However, there are deeper rooted problems:

 - what if messages are reordered (e.g. `mapAsyncUnordered`)? How would that work for slices backed by a shared buffer?
 - what if messages are wrapped, i.e. in a domain specific case class? How would the framework extract the resource to be clean up? And wrapping happens in Akka itself when substreams are used
 - what if messages are *split*, having several classes pointing to the same resource travelling on different parts of a graph? Or the same graph with `mapConcat`? How would we track when to drop the resource, even if we would be able to unwrap?

**Comment by hepin1989** [comment]
@jboner could you do a copy after the message is received from netty?I know with ByteBuf you need call,`ByteBufUtils.release(bytebuf)`.and not forward the `ReferenceContent` message to Akka-Stream.
I don't see a good way to do proper clean up,cause here,we have no clean up hook in akka-stream.

If I am correct,you may want the `filter` do some clean up after the element is filter out.

**Comment by rkuhn** [comment]
@jroper Please refrain from “shouldn’t be a problem” statements. Trust us, we have indeed fought over this and the conclusion is very clear: _it is impossible!_

**Comment by hepin1989** [comment]
I think send the message to dead letter should be impossible,cause there maybe other materializer implement,which is not akka based. a way I can see maybe make the message wrapped by something like:mark the message a `trait CleanupAble{ def cleanup()}` or make the `filter` which accept another argument like `def filter(p: Out ⇒ Boolean)(implicit cleaner:Cleaner[Out]): Repr[Out]`.and the `Cleaner` maybe something like :
```
trait Cleaner[-T]{
def clean(element:T):Unit
}
``` 
so we could have 
```
implicit object ReferenceCountedCleaner extends Cleaner[ReferenceCounted]{
 override def clean(element:ReferenceCounted):Unit = ReferenceCountUtil.release(element)
}
```

**Comment by hepin1989** [comment]
@jroper @rkuhn FYI http://hg.openjdk.java.net/jdk9/jdk9/jdk/rev/7c8488605485

**Comment by hepin1989** [comment]
```
 (e.g. by cleaning them up after a timeout or when their results are observed on the stream output, or by using other means like finalizers etc.).
```
this will delay the cleanup, and causes problems.
